File: Stacks & Queues/stack.py
class Stack:

    def __init__(self) -> None:

        # Create a regular stack
        self.regStack = list()
        self.regSize = 0

        # Create a min stack
        self.minStack = list()

    def pushToMinStack(self, newVal):

        # OBJECTIVE: Add newVal to minStack in ascending order

        # If head element in minStack is bigger than newVal, add it and exit function
        if len(self.minStack) > 0 and self.minStack[-1] >= newVal:
            self.minStack.append(newVal)
            return

        # Create a temporary stack
        tmp = list()

        # Pop head element from minStack until head is bigger than newVal
        while (len(self.minStack) > 0 and self.minStack[-1] < newVal):
            tmp.append(self.minStack.pop())

        # Add newVal to minStack
        self.minStack.append(newVal)

        # Add all elements from tmp stack back to minStack
        while (len(tmp) > 0):
            self.minStack.append(tmp.pop())

    def removeFromMinStack(self, oldVal):

        # OBJECTIVE: Remove oldVal from minStack

        # If head element in minStack is equals to oldVal, remove it and exit function
        if len(self.minStack) > 0 and self.minStack[-1] == oldVal:
            self.minStack.pop()
            return

        # Create a temporary stack
        tmp = list()

        # Pop head element from minStack until oldVal is found
        while (len(self.minStack) > 0 and self.minStack[-1] != oldVal):
            tmp.append(self.minStack.pop())

        # Remove oldVal from minStack
        self.minStack.pop()

        # Add all elements from tmp stack back to minStack
        while (len(tmp) > 0):
            self.minStack.append(tmp.pop())

    def getMin(self):

        # OBJECTIVE: Return minimum value from stack
        return self.minStack[-1]

    def push(self, newVal):

        # OBJECTIVE: Add newVal to top of the stack

        # Add newVal to regular stack
        self.regStack.append(newVal)
        self.regSize += 1

        # Add newVal to minStack and sort it
        self.pushToMinStack(newVal)

    def remove(self):

        # OBJECTIVE: Remove head of stack

        # If stack is empty, exit function
        if self.regSize == 0:
            return None

        # Pop last element
        lastVal = self.regStack.pop()
        self.regSize -= 1

        # Remove element from minStack
        self.removeFromMinStack(lastVal)

        return lastVal

    def peek(self):

        # OBJECTIVE: Return value of top element without removing it from the stack

        # If stack is empty, exit function
        if self.regSize == 0:
            return None

        return self.regStack[-1]

    def isEmpty(self):

        # OBJECTIVE: Return boolean variable representing if stack is empty or not
        return self.regSize == 0

    def printStack(self):

        # OBJECTIVE: Print all elements inside stack from top to bottom

        # NOTE: Since a list is being used with stack operations, I can cycle through the list. Python doesn't have a built-in stack data structure
        ans = ""
        for elem in reversed(self.regStack):
            ans += "{} ".format(elem)

        print(ans)

File: Stacks & Queues/stack.java
// Cracking the Coding Interview - pp. 98 - q 3.2

import java.util.*;

class stack {

    Stack<Integer> regStack;
    int regSize;
    Stack<Integer> minStack;

    public stack() {
        this.regStack = new Stack<Integer>();
        this.regSize = 0;
        this.minStack = new Stack<Integer>();
    }

    private void pushToMinStack(int newVal) {

        // OBJECTIVE: Add a new element to stack in sorted order (ascending order)

        // If head of minStack is bigger than newVal, add newVal to minStack and exit function
        if (minStack.peek() >= newVal) {
            minStack.push(newVal);
            return;
        }

        // Create a temporary stack
        Stack<Integer> tmp = new Stack<Integer>();

        // Continue to pop elements from minValues stack until a larger head is reached
        while (minStack.size() > 0 && minStack.peek() < newVal) {

            // Pop head from minValues and add it to tmp
            tmp.push(minStack.pop());
        }

        // Add newVal to stack
        minStack.push(newVal);

        // Add all elements from tmp stack back to minValue
        while (tmp.size() > 0) {
            minStack.push(tmp.pop());
        }
    }

    private void removeFromMinStack(int oldVal) {

        // OBJECTIVE: Remove oldVal from minStack

        // If head element of minStack equals to oldVal, remove it and exit function
        if (minStack.peek() == oldVal) {
            minStack.pop();
            return;
        }

        // Create a temporary stack
        Stack<Integer> tmp = new Stack<Integer>();

        // Pop an element until oldVal is found in minStack
        while (minStack.size() > 0 && minStack.peek() != oldVal) {
            tmp.push(minStack.pop());
        }

        // Remove head element from stack since it equals to oldVal
        // NOTE: remove() will pop head element from regStack, so we're popping an element that we know it exists inside of minStack
        minStack.pop();

        // Add all elements from tmp stack back to minStack
        while (tmp.size() > 0) {
            minStack.push(tmp.pop());
        }
    }

    public int getMin() {

        // OBJECTIVE: Return smallest value (head element) from stack
        return minStack.peek();
    }

    public void push(int newVal) {

        // OBJECTIVE: Add newVal to the top of the stack
        
        // Add new value to stack
        regStack.add(newVal);
        regSize++;

        // Add new value to array list and sort it
        pushToMinStack(newVal);
    }

    public int remove() {
        
        // OBJECTIVE: Remove the head element of stack and return its value

        // If stack is empty, exit function
        if (regSize == 0) {return -1;}

        // Capture last element
        int lastVal = regStack.pop();
        regSize--;

        // Update minStack
        removeFromMinStack(lastVal);

        return lastVal;
    }

    public int peek() {

        // OBJECTIVE: Return value of head element without removing it

        // If stack is empty, exit function
        if (regSize == 0) {return -1;}

        return regStack.peek();
    }

    public boolean isEmpty() {

        // OBJECTIVE: Return a boolean value representing whether or not stack is empty
        return regSize == 0;
    }
}